### PR TITLE
caddy: `withPlugins` better error messages for untagged plugins

### PR DESCRIPTION
**Pull request by stepbrobd**
## Things done

Context:
https://github.com/NixOS/nixpkgs/pull/358586#issuecomment-2564012300
https://github.com/NixOS/nixpkgs/pull/358586#issuecomment-2564016652

Continuation of #358586, better error messages for untagged plugins

When using `pkgs.caddy.withPlugins`, if a plugin is untagged upstream, users usually provide a commit SHA to replace to the tag/version. This will result in `installCheckPhase` failure as `caddy build-info` output uses [pseudo-version number](https://go.dev/doc/modules/version-numbers#pseudo-version-number) (e.g. `v0.0.0-20170915032832-14c0d48ead0c`).

Example error:

```console
$ nom-build --expr 'with import ./. { }; caddy.withPlugins { plugins = [ "github.com/caddy-dns/powerdns@v1.0.1" "github.com/caddy-dns/cloudflare@89f16b99c18e" "github.com/mholt/caddy-webdav@42168ba04c9dc2cd228ab8c453dbab27654e52e6" ]; hash = "sha256-wqXSd1Ep9TVpQi570TTb96LwzNYvWL5EBJXMJfYWCAk="; }'
...
> Running phase: installCheckPhase
> Plugins not found: github.com/caddy-dns/cloudflare@89f16b99c18e
> Plugins not found: github.com/mholt/caddy-webdav@42168ba04c9dc2cd228ab8c453dbab27654e52e6
...
```

The error output will change to the following:

```console
$ nom-build --expr 'with import ./. { }; caddy.withPlugins { plugins = [ "github.com/caddy-dns/powerdns@v1.0.1" "github.com/caddy-dns/cloudflare@89f16b99c18e" "github.com/mholt/caddy-webdav@42168ba04c9dc2cd228ab8c453dbab27654e52e6" ]; hash = "sha256-wqXSd1Ep9TVpQi570TTb96LwzNYvWL5EBJXMJfYWCAk="; }'
...
> Running phase: installCheckPhase
> Plugin "github.com/caddy-dns/cloudflare" have incorrect tag:
>   specified: "github.com/caddy-dns/cloudflare@89f16b99c18e"
>   got: "github.com/caddy-dns/cloudflare@v0.0.0-20240703190432-89f16b99c18e"
> Plugin "github.com/mholt/caddy-webdav" have incorrect tag:
>   specified: "github.com/mholt/caddy-webdav@42168ba04c9dc2cd228ab8c453dbab27654e52e6"
>   got: "github.com/mholt/caddy-webdav@v0.0.0-20241008162340-42168ba04c9d"
...
```

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ca/caddy/plugins.nix [hunk 1]
```diff
@@ -69,8 +98,18 @@ caddy.overrideAttrs (
...
+        for plugin in "''${notfound[@]}"; do
+          base=''${plugin%@*}
+          specified=''${plugin#*@}
+          while read kind module expected; do
+            [[ "$kind" = "dep" && "$module" = "$base" ]] || continue
+            echo "Plugin \"$base\" have incorrect tag:"
+            echo "  specified: \"$base@$specified\""
+            echo "  got: \"$base@$expected\""
+          done < <($out/bin/caddy build-info)
+        done
```
Comment (vincentbernat) [inline pkgs/by-name/ca/caddy/plugins.nix:112]: Does not display anything if a plugin is not found at all.

Comment (stepbrobd) [reply pkgs/by-name/ca/caddy/plugins.nix:121]: Added a fall back but I don't think this will ever get triggered unless there's smth wrong with xcaddy

**Review by thiagokokada — APPROVED** [review]

File: pkgs/by-name/ca/caddy/plugins.nix [hunk 2]
```diff
@@ -7,19 +8,44 @@
...
+    concatMapStrings
+    elemAt
+    fakeHash
+    filter
+    hasInfix
+    length
+    lessThan
+    sort
+    toShellVar
+    ;
```
Comment (vincentbernat) [inline pkgs/by-name/ca/caddy/plugins.nix:25]: Out of curiosity, is that a best practice? Is it explained somewhere?

Comment (thiagokokada) [reply pkgs/by-name/ca/caddy/plugins.nix:25]: Neither. I particularly don't like but it doesn't hurt (as long you're not using `with lib`).

Comment (stepbrobd) [reply pkgs/by-name/ca/caddy/plugins.nix:25]: I've seen a few modules (in nixpkgs of course) pull in library functions right after getting arguments like these:
- https://github.com/NixOS/nixpkgs/blob/821e42c405b7ff755b0a32f5a5078357fed45354/nixos/modules/services/monitoring/metricbeat.nix#L8-L17
- https://github.com/NixOS/nixpkgs/blob/4506ece030a0c82d078edd0360ea8af6b4d94035/nixos/modules/services/networking/mtr-exporter.nix#L8-L23
- https://github.com/NixOS/nixpkgs/blob/4506ece030a0c82d078edd0360ea8af6b4d94035/nixos/modules/services/networking/ergo.nix#L13-L20
And I've taken quite a liking to this style, so yeah... Not sure if this is the "best practice" tho

**Pull request merged**